### PR TITLE
Implement delay before being able to start table

### LINKED ISSUE #1976: disable start button for 1s after someone joins/leaves
Issue opened by Discord user: Zamiel#8743

### PR DESCRIPTION
**Pull request by timotree3**
Fixes #1976

The UI could use improvement but what do you think about this approach?

### CODE REVIEW

File: server/src/command_table_start.go [hunk 1]
```diff
@@ -65,6 +67,11 @@ func commandTableStart(ctx context.Context, s *Session, d *CommandData) {
 		}
 	}
 
+	if now.Sub(t.DatetimeLastJoined) < time.Second*2 || now.Sub(t.DatetimeLastLeft) < time.Second*2 {
+		s.Warning("Cannot start game immediately after a player joins or leaves. Try again in 2 seconds")
```
Comment (Zamiell) [inline server/src/command_table_start.go:71]: make this match the style of the existing error messages

**Comment by Zamiell** [comment]
suggestion: can we just nuke datetimelastjoined and datetimelastleft and simply use datetimelastaction instead?
(remember to reinitialize it to X during game start)

**Comment by timotree3** [comment]
> suggestion: can we just nuke datetimelastjoined and datetimelastleft and simply use datetimelastaction instead?
> (remember to reinitialize it to X during game start)

I don't think any action should prevent game start. e.g. changing variant shouldn't. Just joining and leaving

**Comment by Zamiell** [comment]
sure, then simply dont have /changevariant affect datetimelastaction

**Comment by timotree3** [comment]
lastaction is currently only used for killing idle tables. I think that concern is unrelated to this PR's and does care about changing variant

File: server/src/command_table_start.go [hunk 2]
```diff
@@ -65,6 +67,11 @@ func commandTableStart(ctx context.Context, s *Session, d *CommandData) {
 		}
 	}
 
+	if now.Sub(t.DatetimeLastJoined) < time.Millisecond*1500 || now.Sub(t.DatetimeLastLeft) < time.Millisecond*1500 {
```
Comment (Zamiell) [inline server/src/command_table_start.go:70]: try this:

```go
recentTimeThreshold := 1.5 * time.Second
playersRecentlyChanged := now.Sub(t.DatetimeLastJoined) < recentTimeThreshold  || now.Sub(t.DatetimeLastLeft) < recentTimeThreshold 
if playersRecentlyChanged  {
```

Comment (timotree3) [reply server/src/command_table_start.go:70]: I doubt you can multiply a duration by a float

Comment (Zamiell) [reply server/src/command_table_start.go]: if you can't multiply a duration by a float, then do:

```go
recentTimeThreshold := 1.5 * time.Millisecond
playersRecentlyChanged := now.Sub(t.DatetimeLastJoined) < recentTimeThreshold || now.Sub(t.DatetimeLastLeft) < recentTimeThreshold 
if playersRecentlyChanged {
```

**Comment by timotree3** [comment]
I was gonna implement graying out the start game button momentarily after a player joins but I don't think that's worthwhile. Especially since in many of these cases the client will learn too late that someone joined. I believe this is ready to merge and we'll get feedback from users as to tweaking to 1.5 seconds constant

**Comment by Zamiell** [comment]
i do think that graying out the start game button would be appropriate.
the avg amount of milliseconds from server to client is 100ms, which is an order of magnitude less than 1.5 seconds

**Comment by timotree3** [comment]
Alright I'm gonna go for a different approach then.

**Comment by Zamiell** [comment]
also i should probably mention that we should put a freeze on any server PRs until channels is merged

**Comment by timotree3** [comment]
I don't think that's necessary. Just merge my changes into your channels branch

**Comment by Zamiell** [comment]
have you seen the code? its like completely different

**Comment by timotree3** [comment]
I haven't

**Pull request closed without merging**